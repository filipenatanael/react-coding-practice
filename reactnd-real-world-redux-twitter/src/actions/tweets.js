export const RECEIVE_TWEETS = 'RECEIVE_TWEETS'

exports function receiveTweets(tweets) {
  return {
    type: RECEIVE_TWEETS,
    tweets
  }
}
