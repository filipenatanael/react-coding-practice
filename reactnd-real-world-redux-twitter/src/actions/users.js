export const RECEIVE_USERS = 'RECEIVE_USERS'

export function receiveUsers(tweets) {
  return {
    type: RECEIVE_USERS,
    tweets
  }
}
