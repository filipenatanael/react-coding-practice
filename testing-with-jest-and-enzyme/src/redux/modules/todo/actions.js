import fetch from 'isomorphic-fetch';

export const fetchRequest = () => ({
  type: 'TODO_REQUEST'
});

export const fetchSuccess = (todos) => ({
  type: 'TODO_SUCCESS',
  todos
});

export const fetchFailure = (failure) => ({
  type: 'TODO_FAILURE',
  failure
});

export const createRequest = () => ({
  type: 'CREATE_REQUEST'
});

export const createSuccess = (todo) => ({
  type: 'CREATE_SUCCESS',
  todo
});

export const createFailure = (failure) => ({
  type: 'CREATE_FAILURE',
  failure
});

export const editRequest = () => ({
  type: 'EDIT_REQUEST'
});

export const editSuccess = (todo) => ({
  type: 'EDIT_SUCCESS',
  todo
});

export const editFailure = (failure) => ({
  type: 'EDIT_FAILURE',
  failure
});

export const deleteRequest = () => ({
  type: 'DELETE_REQUEST'
});

export const deleteSuccess = (id) => ({
  type: 'DELETE_SUCCESS',
  id
});

export const deleteFailure = (failure) => ({
  type: 'DELETE_FAILURE',
  failure
});

export const toggleRequest = () => ({
  type: 'TOGGLE_REQUEST'
});

export const toggleSuccess = (todo) => ({
  type: 'TOGGLE_SUCCESS',
  todo
});

export const toggleFailure = (failure) => ({
  type: 'TOGGLE_FAILURE',
  failure
});

export const fetchTodos = (host) => (dispatch) => {
  let url = '/api/todos';
  url = host ? host + url : url;

  dispatch(fetchRequest());
  return fetch(url)
  .then(response => {
    if (!response.ok) {
      return response.json().then(Promise.reject.bind(Promise));
    }
    return response.json();
  }).then(json => {
    return dispatch(fetchSuccess(json));
  }).catch(err => {
    return dispatch(fetchFailure('Error Fetching Todos!'));
  });
};

export const createTodo = (todo,host) => (dispatch) => {
  let url = '/api/todos';
  url = host ? host + url : url;

  dispatch(createRequest());
  return fetch(url, {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body: JSON.stringify(todo)
    }).then(response => {
      if (!response.ok) {
        return response.json().then(Promise.reject.bind(Promise));
      }
      return response.json();
    }).then(json => {
      return dispatch(createSuccess(json));
    }).catch(err => {
      return dispatch(createFailure('Error Creating Todo!'));
    });
};

export const editTodo = (id, todo,host) => (dispatch) => {
  let url = '/api/todo/' + id;
  url = host ? host + url : url;

  dispatch(editRequest());
  return fetch(url, {
    headers: {
      'Content-Type': 'application/json'
    },
    method: 'PUT',
    body: JSON.stringify(todo)
  }).then(response => {
    if (!response.ok) {
      return response.json().then(Promise.reject.bind(Promise));
    }
    return response.json();
  }).then(json => {
    return dispatch(editSuccess(json));
  }).catch(err => {
    return dispatch(editFailure('Error Editing Todo!'));
  });
}

export const toggleTodo = (id, todo, host) => (dispatch) => {
  let url = '/api/todo/' + id;
  url = host ? host + url : url;

  dispatch(toggleRequest());
  return fetch(url, {
    headers: {
      'Content-Type': 'application/json'
    },
    method: 'PUT',
    body: JSON.stringify(todo)
  }).then(response => {
    if (!response.ok) {
      return response.json().then(Promise.reject.bind(Promise));
    }
    return response.json();
  }).then(json => {
    return dispatch(toggleSuccess(json));
  }).catch(err => {
    return dispatch(toggleFailure('Error Updating Complete Status!'));
  });
}

export const deleteTodo = (id, host) => (dispatch) => {
  let url = '/api/todo/' + id;
  url = host ? host + url : url;

  dispatch(deleteRequest());
  return fetch(url, {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'DELETE',
    }).then(response => {
      if (!response.ok) {
        return response.json().then(Promise.reject.bind(Promise));
      }
      return response.json();
    }).then(json => {
      return dispatch(deleteSuccess(id));
    }).catch(err => {
      return dispatch(deleteFailure('Error Deleting Todo!'));
    });
}
