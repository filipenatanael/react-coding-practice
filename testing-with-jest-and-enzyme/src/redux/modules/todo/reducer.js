const updateTodos = (todos, todo) => todos.map(t => t.id === todo.id ? todo : t)
const removeTodo = (todos, id) => todos.filter(t => t.id !== id)

export const initialState = { // Exporting it for test purposes
  requesting: false,
  todos: [],
  error: null,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TODO_REQUEST':
      return Object.assign({}, state,
        {
          requesting: true,
          error: null,
        }
      );
    case 'TODO_SUCCESS':
      return Object.assign({}, state,
        {
          requesting: false,
          todos: action.todos
        }
      );
    case 'TODO_FAILURE':
    case 'CREATE_FAILURE':
    case 'EDIT_FAILURE':
    case 'TOGGLE_FAILURE':
    case 'DELETE_FAILURE':
      return Object.assign({}, state,
        {
          requesting: false,
          error: action.failure
        }
      );
    case 'CREATE_REQUEST':
      return Object.assign({}, state,
        {
          requesting: true,
          error: null,
        }
      );
    case 'CREATE_SUCCESS':
      return Object.assign({}, state,
        {
          requesting: false,
          todos: [action.todo, ...state.todos]
          // todos: true // Write code for this
        }
      );

    case 'EDIT_REQUEST':
      return Object.assign({}, state,
        {
          requesting: true,
          error: null,
        }
      );
    case 'EDIT_SUCCESS':
    case 'TOGGLE_SUCCESS':
      return Object.assign({}, state,
        {
          requesting: false,
          todos: updateTodos(state.todos, action.todo)
        }
      );
    case 'TOGGLE_REQUEST':
      return Object.assign({}, state,
        {
          requesting: true,
          error: null
        }
      );
    case 'DELETE_REQUEST':
      return Object.assign({}, state,
        {
          requesting: true,
          error: null
        }
      );
    case 'DELETE_SUCCESS':
      return Object.assign({}, state,
        {
          requesting: false,
          todos: removeTodo(state.todos, action.id)
        }
      );
    default:
      return state;
  }
}
