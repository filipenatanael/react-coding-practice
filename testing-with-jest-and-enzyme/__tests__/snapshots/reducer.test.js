import reducer from '../../src/redux/modules/todo/reducer'
import * as actions from '../../src/redux/modules/todo/actions'

describe('todos reducer', () => {
  let todo = {
    "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
    "title": "Todo",
    "project": "Project",
    "done": false,
    "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
    "createdAt": "2017-03-02T23:04:38.003Z",
    "modifiedAt": "2017-03-22T16:44:29.034Z"
  }
  let todos = [
    {
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "title": "Todo",
      "project": "Project",
      "done": false,
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "createdAt": "2017-03-02T23:04:38.003Z",
      "modifiedAt": "2017-03-22T16:44:29.034Z"
    },
    {
      "id": "a9e01001-3750-4c88-9c67-ba3e68255317",
      "title": "New Todo",
      "project": "New Project",
      "done": true,
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image3.jpg",
      "createdAt": "2017-03-18T06:45:57.337Z",
      "modifiedAt": "2017-03-18T17:15:23.996Z"
    }
  ]
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('should handle TODO_REQUEST', () => {
    expect(
      reducer({
        requesting: false,
        todos: [],
        error: null
      },
      {
        type: 'TODO_REQUEST'
      })
    ).toMatchSnapshot()
  })

  it('should handle TODO_SUCCESS', () => {
    expect(
      reducer({
        requesting: true,
        todos: [],
        error: null
      },
      {
        type: 'TODO_SUCCESS',
        todos
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle TODO_FAILURE', () => {
    expect(
      reducer({
        requesting: true,
        todos: [],
        error: null
      },
      {
        type: 'TODO_FAILURE',
        failure: 'Error Fetching Todos!'
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle CREATE_REQUEST', () => {
    expect(
      reducer({
        requesting: false,
        todos: [],
        error: null
      },
      {
        type: 'CREATE_REQUEST'
      })
    ).toMatchSnapshot()
  })

  it('should handle CREATE_SUCCESS', () => {
    expect(
      reducer({
        requesting: true,
        todos: [],
        error: null
      },
      {
        type: 'CREATE_SUCCESS',
        todo,
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle CREATE_FAILURE', () => {
    expect(
      reducer({
        requesting: true,
        todos: [],
        error: null
      },
      {
        type: 'CREATE_FAILURE',
        failure: 'Error Creating Todo!'
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle EDIT_REQUEST', () => {
    expect(
      reducer({
        requesting: false,
        todos: [todo],
        error: null
      },
      {
        type: 'EDIT_REQUEST'
      })
    ).toMatchSnapshot()
  })

  it('should handle EDIT_SUCCESS', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'EDIT_SUCCESS',
        todo: Object.assign({}, todo, {
            "title": "Todo Edited"
          })
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle EDIT_FAILURE', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'EDIT_FAILURE',
        failure: 'Error Editing Todo!'
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle TOGGLE_REQUEST', () => {
    expect(
      reducer({
        requesting: false,
        todos: [todo],
        error: null
      },
      {
        type: 'TOGGLE_REQUEST'
      })
    ).toMatchSnapshot()
  })

  it('should handle TOGGLE_SUCCESS', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'TOGGLE_SUCCESS',
        todo: Object.assign({},todo,{
            "done": true,
          })
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle TOGGLE_FAILURE', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'TOGGLE_FAILURE',
        failure: 'Error Updating Complete Status!'
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle DELETE_REQUEST', () => {
    expect(
      reducer({
        requesting: false,
        todos: [todo],
        error: null
      },
      {
        type: 'DELETE_REQUEST'
      })
    ).toMatchSnapshot()
  })

  it('should handle DELETE_SUCCESS', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'DELETE_SUCCESS',
        id: "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8"
      }
      )
    ).toMatchSnapshot()
  })

  it('should handle DELETE_FAILURE', () => {
    expect(
      reducer({
        requesting: true,
        todos: [todo],
        error: null
      },
      {
        type: 'DELETE_FAILURE',
        failure: 'Error Deleting Todo!'
      }
      )
    ).toMatchSnapshot()
  })
})