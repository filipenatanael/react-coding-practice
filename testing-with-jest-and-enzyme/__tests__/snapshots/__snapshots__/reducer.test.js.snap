// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducer should handle CREATE_FAILURE 1`] = `
Object {
  "error": "Error Creating Todo!",
  "requesting": false,
  "todos": Array [],
}
`;

exports[`todos reducer should handle CREATE_REQUEST 1`] = `
Object {
  "error": null,
  "requesting": true,
  "todos": Array [],
}
`;

exports[`todos reducer should handle CREATE_SUCCESS 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle DELETE_FAILURE 1`] = `
Object {
  "error": "Error Deleting Todo!",
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle DELETE_REQUEST 1`] = `
Object {
  "error": null,
  "requesting": true,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle DELETE_SUCCESS 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [],
}
`;

exports[`todos reducer should handle EDIT_FAILURE 1`] = `
Object {
  "error": "Error Editing Todo!",
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle EDIT_REQUEST 1`] = `
Object {
  "error": null,
  "requesting": true,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle EDIT_SUCCESS 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo Edited",
    },
  ],
}
`;

exports[`todos reducer should handle TODO_FAILURE 1`] = `
Object {
  "error": "Error Fetching Todos!",
  "requesting": false,
  "todos": Array [],
}
`;

exports[`todos reducer should handle TODO_REQUEST 1`] = `
Object {
  "error": null,
  "requesting": true,
  "todos": Array [],
}
`;

exports[`todos reducer should handle TODO_SUCCESS 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
    Object {
      "createdAt": "2017-03-18T06:45:57.337Z",
      "done": true,
      "id": "a9e01001-3750-4c88-9c67-ba3e68255317",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image3.jpg",
      "modifiedAt": "2017-03-18T17:15:23.996Z",
      "project": "New Project",
      "title": "New Todo",
    },
  ],
}
`;

exports[`todos reducer should handle TOGGLE_FAILURE 1`] = `
Object {
  "error": "Error Updating Complete Status!",
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle TOGGLE_REQUEST 1`] = `
Object {
  "error": null,
  "requesting": true,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": false,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should handle TOGGLE_SUCCESS 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [
    Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "done": true,
      "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
      "image": "https://raw.githubusercontent.com/andela-jkithome/image_files/master/images/image5.jpg",
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Project",
      "title": "Todo",
    },
  ],
}
`;

exports[`todos reducer should return the initial state 1`] = `
Object {
  "error": null,
  "requesting": false,
  "todos": Array [],
}
`;
